fusillade-api.yml

```diff
@@ -4,164 +4,465 @@ info:
+                      type: array
+                      items:
+                        type: string
+        404:
+          description: Resource not found.
         default:
           description: Unexpected error
-  /groups:
     put:
-      summary: Create or update a group and attached policy.
+      summary: Add user to groups
       security:
-        - openId:
-            - https://www.googleapis.com/auth/userinfo.email
+        - bearerAuth: []
       parameters:
-        - name: json_request_body
-          in: body
-          required: true
-          schema:
-            type: object
-            properties:
-              policy:
-                description: IAM policy to attach, if any
-                type: object
+        - $ref: '#/components/parameters/user_id'
+        - $ref: '#/components/parameters/action'
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                groups:
+                  type: array
+                  items:
+                    type: string
+      responses:
+        200:
+          description: User added to groups
+        401:
+          description: The user is not authorized because the object has already been created and the user does not have permission to modify, or user does not have permission to create objects.
+        409:
+          description: The user is already apart of the group
+        default:
+          description: Unexpected error
+  /users/{user_id}/roles:
+    get:
+      summary: Retrieve roles a user is in.
+      security:
+        - bearerAuth: []
+      parameters:
+        - $ref: '#/components/parameters/user_id'
       responses:
-        201:
+        200:
           description: OK
-          schema:
-            type: object
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  - role:
+                      type: array
+                      items:
+                        type: string
+        404:
+          description: Resource not found.
+        default:
+          description: Unexpected error
+    put:
+      summary: Add user to roles
+      security:
+        - bearerAuth: []
+      parameters:
+        - $ref: '#/components/parameters/user_id'
+        - $ref: '#/components/parameters/action'
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                roles:
+                  type: array
+                  items:
+                    type: string
+      responses:
+        200:
+          description: User added to roles
+        401:
+          description: The user is not authorized because the object has already been created and the user does not have permission to modify, or user does not have permission to create objects.
+        409:
```

Comment:
If PUT is supposed to be idempotent, is there a reason to return a 409 if the user is already a part of the role instead of a 304?

Reply:
changed to 304